mod/state-transition/pkg/core/state_processor_validators.go

```diff
@@ -58,4 +60,45 @@ func (sp *StateProcessor[
 			}, nil
 		},
 	)
+	if err != nil {
+		return nil, err
+	}
+
+	res := make([]*transition.ValidatorUpdate, 0)
+
+	prevValsSet := make(map[string]math.Gwei, len(sp.prevEpochValidators))
+	for _, v := range sp.prevEpochValidators {
+		prevValsSet[string(v.Pubkey[:])] = v.EffectiveBalance
+	}
+
+	for _, newVal := range currentValSet {
+		key := string(newVal.Pubkey[:])
+		oldBal, found := prevValsSet[key]
+		if !found {
+			// new validator, we add it with its weight
+			res = append(res, newVal)
+			continue
+		}
+		if oldBal != newVal.EffectiveBalance {
+			// validator updated, we add it with new weight
+			res = append(res, newVal)
+		}
+
+		// consume pre-existing validators
+		delete(prevValsSet, key)
+	}
+
+	// prevValsSet now contains all evicted validators (and only those)
+	for pkBytes := range prevValsSet {
+		//#nosec:G703 // bytes comes from a pk
+		pk, _ := bytes.ToBytes48([]byte(pkBytes))
+		res = append(res, &transition.ValidatorUpdate{
+			Pubkey:           pk,
+			EffectiveBalance: 0, // signal val eviction to consensus
+		})
+	}
```

Comment:
we do not evict validators yet, so this should not really happen. Still added for completeness

Reply:
in fact we may evict them as soon as we introduce a cap on the validators set